### PR TITLE
Shifted Composite L2 Norm

### PR DESCRIPTION
**Pull request by MaxenceGollier**
Implementation of the proximal operator of a composite term $||Ax+b||_2$. This is different from other operators of this library in the fact that shifts are no longer simply $\psi(x+s)$ for some non-differentiable function $\psi$. 
Hence, some abstract types are added in ShiftedProximalOperators.jl. 

In practice, such proximal operators are useful when we want to make a model of a penalty term $x \xrightarrow{} ||c(x)||_2$. The first-order model of the penalty term is then $s \xrightarrow{} ||c(x) + J(x)s||_2$.

As mentionned above, the shift is no longer simply $\psi(x+s)$. Hence, we add a CompositeNormL2 struct which mimics the unshifted function and implements $x\xrightarrow{} ||c(x)||_2$. On shifts, this transforms to a ShiftedCompositeNormL2 struct which implements $s\xrightarrow{} ||c(x)+J(x)s||_2$.

### CODE REVIEW

File: src/shiftedCompositeNormL2.jl [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+    if nnz(R1[rp[i],cp[:]]) != 0
+      push!(rows_R, i)
+    end
+  end
+  for i = 1 : size(R1, 2)
+    if abs(R1[rp[rows_R[i]], cp[i]]) < eps(R)^0.75
+      full_row_rank = false
+      break
+    end
+  end
```
Comment (dpo) [inline src/shiftedCompositeNormL2.jl:158]: I don't think you need to copy the diagonal of R to check its values, do you? You could simply scan the sparsity structure and check the value of elements on the diagonal?!

File: src/shiftedCompositeNormL2.jl [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+
+  spmat = qrm_spmat_init(ψ.Aᵧ; sym=false) # TODO: preallocate this
+  spfct = qrm_spfct_init(spmat) # TODO: preallocate this
+  qrm_set(spfct, "qrm_keeph", 0)
+
+  # Check interior convergence
+  qrm_analyse!(spmat, spfct; transp='t')
+  qrm_factorize!(spmat, spfct, transp='t')
+
+  qrm_solve!(spfct, ψ.g, ψ.p, transp='t')
```
Comment (dpo) [inline src/shiftedCompositeNormL2.jl:124]: Shouldn't you check the rank of R before solving?

Comment (dpo) [reply src/shiftedCompositeNormL2.jl:124]: Also, the code is a bit hard to follow because the notation does not match the paper. There's p here and p in the paper, but it's not the same p.

File: src/shiftedCompositeNormL2.jl [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+  qrm_solve!(spfct, ψ.p, ψ.sol, transp='n')
+
+  # 1 Step of iterative refinement
```
Comment (dpo) [inline src/shiftedCompositeNormL2.jl:127]: Let's include this directly into QRMumps, so we don't have to do it explicitly.

File: src/shiftedCompositeNormL2.jl [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+      full_row_rank = false
+      break
+    end
+  end
+
+  if !full_row_rank 
+
+    γ = eps(R)
+    
+    qrm_update!(spmat,[ψ.A.vals; fill(eltype(ψ.A.vals)(sqrt(γ)),ψ.A.m)])
```
Comment (dpo) [inline src/shiftedCompositeNormL2.jl:164]: `sqrt(γ)` should already have the right type here, but we should be careful because QRMumps doesn't accept arbitrary types.

File: src/shiftedCompositeNormL2.jl [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+
+  mul!(ψ.dp, ψ.Aᵧ', ψ.sol)
+  mul!(ψ.dsol, ψ.Aᵧ, ψ.dp)
+
+  ψ.res .-= ψ.dsol
+  if norm(ψ.res) > eps(R)^0.75
+    qrm_solve!(spfct, ψ.res, ψ.dp, transp='t')
+    qrm_solve!(spfct, ψ.dp, ψ.dsol, transp='n')
+    ψ.sol .+= ψ.dsol
+  end
```
Comment (dpo) [inline src/shiftedCompositeNormL2.jl:138]: Why not call `_iterative_refinement!()` here?

File: src/shiftedCompositeNormL2.jl [hunk 6]
```diff
@@ -0,0 +1,228 @@
...
+  if !full_row_rank 
+
+    γ = eps(R)
+    
+    qrm_update!(spmat,[ψ.A.vals; fill(eltype(ψ.A.vals)(sqrt(γ)),ψ.A.m)])
+    qrm_factorize!(spmat, spfct, transp='t')
+
+    qrm_solve!(spfct, ψ.g, ψ.p, transp='t')
+    qrm_solve!(spfct, ψ.p, ψ.sol, transp='n')
+    qrm_solve!(spfct, ψ.sol, ψ.p, transp='t')
```
Comment (dpo) [inline src/shiftedCompositeNormL2.jl:169]: Introduce intermediate functions to perform the solve so we don't repeat code.

File: src/shiftedCompositeNormL2.jl [hunk 7]
```diff
@@ -0,0 +1,228 @@
...
+    _iterative_refinement!(spfct, ψ)
+
+    ψ.sol .*= -1
```
Comment (dpo) [inline src/shiftedCompositeNormL2.jl:172]: Technically, p has the wrong sign here, but I guess you only need its norm, so it's ok. Please add a comment to clarify this.

File: src/shiftedCompositeNormL2.jl [hunk 8]
```diff
@@ -0,0 +1,228 @@
...
+
+    if norm(ψ.sol) ≤ σ*ψ.h.lambda + eps(R) # We are in the hard-case, we consider ψ.sol is a good approximation of the least-norm solution
```
Comment (dpo) [inline src/shiftedCompositeNormL2.jl:174]: This isn't the hard case. The hard case doesn't happen in convex problems.

File: src/shiftedCompositeNormL2.jl [hunk 9]
```diff
@@ -0,0 +1,228 @@
...
+      mul!(y, ψ.A', ψ.sol)
+      y .+= q
```
Comment (dpo) [inline src/shiftedCompositeNormL2.jl:176]: Here, something happened with a - sign. I think it's because the right-hand side had the wrong sign in the system with $R^T R$ that you solved above. Please explain in comments.

File: src/shiftedCompositeNormL2.jl [hunk 10]
```diff
@@ -0,0 +1,228 @@
...
+
+  end
+  
+  # Scalar Root finding
+  k = 0
+  elapsed_time = time() - start_time
+  γ₊ = γ 
+  if norm(ψ.sol) > σ*ψ.h.lambda
+
+    while abs(norm(ψ.sol) - σ*ψ.h.lambda) > eps(R)^0.75 && k < max_iter && elapsed_time < max_time
```
Comment (dpo) [inline src/shiftedCompositeNormL2.jl:188]: It would be safer to write `norm(ψ.sol) > σ*ψ.h.lambda + ...`

File: src/shiftedCompositeNormL2.jl [hunk 11]
```diff
@@ -0,0 +1,228 @@
...
+
+      γ₊ += (norm(ψ.sol)/(σ*ψ.h.lambda) - 1.0)*(norm(ψ.sol)/norm(ψ.p))^2
```
Comment (dpo) [inline src/shiftedCompositeNormL2.jl:190]:
```suggestion
      solNorm = norm(ψ.sol)
      γ₊ += (solNorm / (σ * ψ.h.lambda) - 1) * (solNorm / norm(ψ.p))^2
```

File: src/shiftedCompositeNormL2.jl [hunk 12]
```diff
@@ -0,0 +1,228 @@
...
+      γ = γ₊ > 0 ? γ₊ : θ*γ
```
Comment (dpo) [inline src/shiftedCompositeNormL2.jl:191]: If `γ₊ ≤ 0`, then `γ ≤ 0` too. This isn't right. You need to reinitialize `γ` to a positive value.

**Review by dpo — CHANGES_REQUESTED** [review]

File: Project.toml [hunk 13]
```diff
@@ -4,15 +4,23 @@ authors = ["rjbaraldi <rbaraldi@uw.edu>"]
...
 
 [deps]
+ADNLPModels = "54578032-b7ea-4c30-94aa-7cbd1cce6c9a"
 LinearAlgebra = "37e2e46d-f89d-539d-b4ee-838fcccc9c8e"
+NLPModels = "a4795742-8479-5a88-8948-cc11e1c8c1a6"
 OpenBLAS32_jll = "656ef2d0-ae68-5445-9ca0-591084a874a2"
+OptimizationProblems = "5049e819-d29b-5fba-b941-0eee7e64c1c6"
 ProximalOperators = "a725b495-10eb-56fe-b38b-717eba820537"
+QRMumps = "422b30a1-cc69-4d85-abe7-cc07b540c444"
+RegularizedProblems = "ea076b23-609f-44d2-bb12-a4ae45328278"
```
[the remark this answers is not in the document]
Comment (MaxenceGollier) [reply Project.toml:14]: QRMumps is required in the prox computation.

File: src/shiftedCompositeNormL2.jl (hunk 1 above)
Comment (MaxenceGollier) [reply src/shiftedCompositeNormL2.jl]: I don't understand, what do you mean by scanning the sparsity ?

File: src/shiftedCompositeNormL2.jl [hunk 14]
```diff
@@ -0,0 +1,223 @@
...
+  ShiftedCompositeNormL2(ψ.h.lambda, ψ.c!, ψ.J!, A, b)
+end
+ 
+shifted(
+  ψ::CompositeNormL2{R, V0, V1, V2, V3},
+  xk::AbstractVector{R}
+) where {R <: Real, V0 <: Function, V1 <: Function, V2 <: AbstractMatrix{R}, V3 <: AbstractVector{R}} = begin
+  b = similar(ψ.b)
+  ψ.c!(b,xk)
+  A = copy(ψ.A)
```
[the remark this answers is not in the document]
Comment (MaxenceGollier) [reply src/shiftedCompositeNormL2.jl:87]: You are right. I think it was because we did not have a `similar` function for `SparseMatrixCOO` when I wrote this. This means the compatibility of SparseMatrixCOO in `Projet.toml` should be updated as well to match the PR [SparseMatricesCOO.jl#40](https://github.com/JuliaSmoothOptimizers/SparseMatricesCOO.jl/pull/40)

File: src/shiftedCompositeNormL2.jl (hunk 2 above)
Comment (MaxenceGollier) [reply src/shiftedCompositeNormL2.jl:128]: It is to reduce the number of allocations. But we may change it if you find readibility more important: 
We need to solve 
```math
R^T R R^T p = g
```
For this, we could allocate three vectors $p_1, p_2,p$ and solve 
```math
\begin{align}
R^T p_1 &= g \\
R p_2 &= p_1\\
R^T p &= p_2
\end{align}
```
What I do instead is
```math
\begin{align}
R^T p &= g \\
R p_2 &= p\\
R^T p &= p_2
\end{align}
```
Where in fact $p_2$ is $q_k$ with the notation from the paper (In the code, it is called `ψ.sol` but I can rename it). This allows to save one allocation...

File: src/shiftedCompositeNormL2.jl (hunk 3 above)
Comment (MaxenceGollier) [reply src/shiftedCompositeNormL2.jl]: I will open a PR

File: src/shiftedCompositeNormL2.jl [hunk 15]
```diff
@@ -0,0 +1,223 @@
...
+fun_name(ψ::ShiftedCompositeNormL2) = "shifted L2 norm"
+fun_expr(ψ::ShiftedCompositeNormL2) = "t ↦ ‖c(xk) + J(xk)t‖₂"
+fun_params(ψ::ShiftedCompositeNormL2) = "c(xk) = $(ψ.b)\n" * " "^14 * "J(xk) = $(ψ.A)\n"
+
+function prox!(
+  y::AbstractVector{R},
+  ψ::ShiftedCompositeNormL2{R, V0, V1, V2, V3},
+  q::AbstractVector{R},
+  σ::R;
+  maxiter = 10000
```
[the remark this answers is not in the document]
Comment (MaxenceGollier) [reply src/shiftedCompositeNormL2.jl]: Golub-Riley is not implemented yet.

**Comment by dpo** [comment]
@MaxenceGollier It looks like all tests fail here.

**Comment by dpo** [comment]
@MaxenceGollier Could you please prepare a new version of QRMumps and update this branch?

File: src/shiftedCompositeNormL2.jl [hunk 16]
```diff
@@ -0,0 +1,173 @@
...
+    # The solution is not α = 0, prepare root finding
+    qrm_update_shift_spmat!(shifted_spmat, α)
+    _obj_dot_grad!(spmat, spfct, ψ.p, ψ.q, ψ.g, ψ.dq)
+  end
+  
+  # Scalar Root finding
+  k = 0
+  elapsed_time = time() - start_time
+  α₊ = α 
+  if norm(ψ.q) > ν*ψ.h.lambda
```
Comment (dpo) [inline src/shiftedCompositeNormL2.jl:141]: Above, you add `eps(R)` to the rhs. Why not here?

File: src/shiftedCompositeNormL2.jl [hunk 17]
```diff
@@ -0,0 +1,173 @@
...
+
+  # Check interior convergence
+  qrm_analyse!(spmat, spfct; transp='t')
+  _obj_dot_grad!(spmat, spfct, ψ.p, ψ.q, ψ.g, ψ.dq)
+  
+  # Check full-rankness
+  full_row_rank = (qrm_get(spfct,"qrm_rd_num") == 0)
+  if !full_row_rank
+    α = αmin
+    qrm_golub_riley!(ψ.shifted_spmat, spfct, ψ.p, ψ.g, ψ.dp, ψ.q, ψ.dq, transp = 't', α = α, tol = eps(R)^(0.75))
```
Comment (dpo) [inline src/shiftedCompositeNormL2.jl:119]: These long and complicated function calls make it difficult to remember what each variable contains. Could you add comments to explain what `ψ.p`, etc., contain?

File: src/shiftedCompositeNormL2.jl [hunk 18]
```diff
@@ -0,0 +1,173 @@
...
+
+    # Compute residual
+    qrm_spmat_mv!(spmat, R(1), ψ.q, R(0), ψ.dp, transp = 't')
+    qrm_spmat_mv!(spmat, R(1), ψ.dp, R(0), ψ.dq, transp = 'n')
+    @. ψ.dq = ψ.dq - ψ.g
```
Comment (dpo) [inline src/shiftedCompositeNormL2.jl:124]: Same comment here.

File: src/shiftedCompositeNormL2.jl [hunk 19]
```diff
@@ -0,0 +1,173 @@
...
+  end
+  
+  # Scalar Root finding
+  k = 0
+  elapsed_time = time() - start_time
+  α₊ = α 
+  if norm(ψ.q) > ν*ψ.h.lambda
+    while abs(norm(ψ.q) - ν*ψ.h.lambda) > eps(R)^0.3 && k < max_iter && elapsed_time < max_time
+
+      solNorm = norm(ψ.q)
```
Comment (dpo) [inline src/shiftedCompositeNormL2.jl:144]: you’ve computed `norm(q)` three times already.

File: src/shiftedCompositeNormL2.jl [hunk 20]
```diff
@@ -0,0 +1,173 @@
...
+    qrm_update_shift_spmat!(shifted_spmat, α)
+    _obj_dot_grad!(spmat, spfct, ψ.p, ψ.q, ψ.g, ψ.dq)
+  end
+  
+  # Scalar Root finding
+  k = 0
+  elapsed_time = time() - start_time
+  α₊ = α 
+  if norm(ψ.q) > ν*ψ.h.lambda
+    while abs(norm(ψ.q) - ν*ψ.h.lambda) > eps(R)^0.3 && k < max_iter && elapsed_time < max_time
```
[the remark this answers is not in the document]
Comment (MaxenceGollier) [reply src/shiftedCompositeNormL2.jl:142]: I don't understand your comment

[the remark this answers is not in the document]
Comment (dpo) [reply src/shiftedCompositeNormL2.jl]: On line 128, you check
```julia
norm(ψ.q) ≤ ν*ψ.h.lambda + eps(T)
```
Then on line 141,
```julia
if norm(ψ.q) > ν*ψ.h.lambda
```
and on line 142,
```julia
abs(norm(ψ.q) - ν*ψ.h.lambda) > eps(R)^0.3
```
If you don’t use the same tolerance everywhere, it’s possible (in theory) to not satisfy any of those tests.

File: src/shiftedCompositeNormL2.jl [hunk 21]
```diff
@@ -0,0 +1,177 @@
...
+  # Scalar Root finding
+  k = 0
+  elapsed_time = time() - start_time
+  α₊ = α 
+
+  norm_q = norm(ψ.q)
+  if abs(norm_q - ν*ψ.h.lambda) > atol
+    while abs(norm_q - ν*ψ.h.lambda) > atol && k < max_iter && elapsed_time < max_time
+
+      α₊ += (norm_q / (ν * ψ.h.lambda) - 1) * (norm_q / norm(ψ.p))^2
```
Comment (dpo) [inline src/shiftedCompositeNormL2.jl:148]:
```suggestion
  νλ = ν * ψ.h.lambda
  if abs(norm_q - νλ) > atol
    while abs(norm_q - νλ) > atol && k < max_iter && elapsed_time < max_time

      α₊ += (norm_q / (νλ - 1) * (norm_q / norm(ψ.p))^2
```

File: src/shiftedCompositeNormL2.jl [hunk 22]
```diff
@@ -0,0 +1,178 @@
...
+  k = 0
+  elapsed_time = time() - start_time
+  α₊ = α 
+
+  norm_q = norm(ψ.q)
+  νλ = ν * ψ.h.lambda
+  if abs(norm_q - νλ) > atol
+    while abs(norm_q - νλ) > atol && k < max_iter && elapsed_time < max_time
+
+      α₊ += (norm_q / (νλ - 1) * (norm_q / norm(ψ.p))^2
```
Comment (dpo) [inline src/shiftedCompositeNormL2.jl:149]:
```suggestion
      α₊ += (norm_q / νλ - 1) * (norm_q / norm(ψ.p))^2
```
